codejail_service/codejail.py

```diff
@@ -29,12 +33,30 @@ def safe_exec(code, input_globals, **kwargs):
     # fixing this.
 
     # pylint: disable=import-outside-toplevel
+    from codejail.safe_exec import SafeExecException
     from codejail.safe_exec import safe_exec as real_safe_exec
 
     # Prevent mutation of input
     output_globals = deepcopy(input_globals)
     try:
         real_safe_exec(code, output_globals, **kwargs)
         return (output_globals, None)
-    except BaseException as e:
+    except SafeExecException as e:
+        # These exception messages can be safely returned to the user, as they
+        # indicate an error encountered by the jailed subprocess.
         return (output_globals, str(e))
+    except JSONDecodeError as e:  # pragma: no cover
+        # codejail communicates with the jailed subprocess via JSON on stdout;
+        # the jailed code can corrupt the JSON (accidentally or on purpose). If
+        # this happens, we can just log that it happened, and we don't need the
+        # full traceback.
+        #
+        # A normal print() shouldn't cause this, but forking the process can,
+        # because both processes then try to write to stdout.
+        log.warning(f"Corrupted JSON from jailed process: {e!r}")
+        return (output_globals, "Sandboxed code produced corrupted stdout.")
+    except BaseException as e:  # pragma: no cover
+        # Don't give details of unexpected exception, as it may indicate a bug
+        # in the codejail service rather than the jailed code.
+        log.warning(f"Unexpected error type from safe_exec: {e!r}", exc_info=True)
```

Comment:
Given that this is unexpected, we don't know if warning or error would be more appropriate, so I'd probably err toward error so this stands out and maybe we'd look into this. It also make me wonder where it might make sense to use edx-django-utils `record_exception` to capture these caught exceptions.